mirgecom/logging_quantities.py

```diff
@@ -0,0 +1,318 @@
+
+        from functools import partial
+
+        if op == "min":
+            self._discr_reduction = partial(self.discr.nodal_min, "vol")
+            self.rank_aggr = min
+        elif op == "max":
+            self._discr_reduction = partial(self.discr.nodal_max, "vol")
+            self.rank_aggr = max
+        elif op == "norm":
+            self._discr_reduction = partial(self.discr.norm)
+            self.rank_aggr = max
+        else:
+            raise ValueError(f"unknown operation {op}")
+
+    @property
+    def default_aggregator(self):
+        """Rank aggregator to use."""
+        return self.rank_aggr
+
+    def __call__(self):
+        """Return the requested quantity."""
+        if self.state_vars is None:
+            return None
+
+        quantity = self.state_vars[self.quantity]
+
+        if self.axis is not None:  # momentum
+            return self._discr_reduction(quantity[self.axis])
+        else:  # mass, energy
+            return self._discr_reduction(quantity)
+
+# }}}
+
+
+# {{{ Kernel profile quantities
+
+class KernelProfile(MultiLogQuantity):
+    """Logging support for statistics of the OpenCL kernel profiling (time, \
+    num_calls, flops, bytes_accessed, footprint).
+
+    Parameters
+    ----------
+    actx
+        The array context from which to collect statistics. Must have profiling
+        enabled in the OpenCL command queue.
+
+    kernel_name
+        Name of the kernel to profile.
+    """
+
+    def __init__(self, actx: PyOpenCLArrayContext,
+                 kernel_name: str) -> None:
+        from mirgecom.profiling import PyOpenCLProfilingArrayContext
+        assert isinstance(actx, PyOpenCLProfilingArrayContext)
+
+        units = ["s", "GFlops", "1", "GByte", "GByte"]
+
+        from dataclasses import fields
+        from mirgecom.profiling import ProfileResultsForKernel
+
+        names = []
+        units = []
+
+        for f in fields(ProfileResultsForKernel):
+            names.append(f"{kernel_name}_{f.name}")
+            if f.name == "num_calls":
+                units.append("1")
+            elif f.name == "flops":
+                units.append("GFlops")
+            elif f.name == "time":
+                units.append("s")
+            elif f.name == "bytes_accessed" or f.name == "footprint_bytes":
+                units.append("GByte")
+            else:
+                raise RuntimeError(f"unknown unit for field {f.name}")
+
+        super().__init__(names, units)
+
+        self.kernel_name = kernel_name
+        self.actx = actx
+
+    def __call__(self) -> list:
+        """Return the requested kernel profile quantity."""
+        from dataclasses import astuple
+        r = self.actx.get_profiling_data_for_kernel(self.kernel_name)
+        return astuple(r)
+
+# }}}
+
+
+# {{{ Memory profiling
+
+class PythonMemoryUsage(LogQuantity):
+    """Logging support for Python memory usage (RSS, host).
+
+    Uses :mod:`psutil` to track memory usage. Virtually no overhead.
```

Comment:
This is missing from the docs. (Otherwise the unresolved reference to the `psutil` docs would cause the doc build to fail.)

Reply:
Added